### PR TITLE
implements a simple import table for the ELF loader

### PR DESCRIPTION
**Pull request by m4b**
1. Updates to goblin 0.0.8 (mostly bug fixes w.r.t. relocation values, and better relocation handling, and also removes all `method` accesses on the elf field members, in addition to using plain vectors instead of the hacks before, so should be much much nicer api to use
2. More importantly, adds some basic infrastructure for resolving symbol imports to their names.

A new field `imports` is added to the `Project` struct, which is a hashmap from `u64 -> String`, e.g., the import address and the resolved name.

Some sample output on a binary with `puts`, `printf` imported from `libc`:

```
DEBUG:panopticon::loader: Symbol:  @ 0x0: st_name: 0 LOCAL NOTYPE st_other: 0 st_shndx: 0 st_value: 0 st_size: 0
DEBUG:panopticon::loader: Symbol: puts @ 0x0: st_name: 11 GLOBAL FUNC st_other: 0 st_shndx: 0 st_value: 0 st_size: 0
DEBUG:panopticon::loader: Import match puts: 0x601018 st_name: 11 GLOBAL FUNC st_other: 0 st_shndx: 0 st_value: 0 st_size: 0
DEBUG:panopticon::loader: Symbol: printf @ 0x0: st_name: 16 GLOBAL FUNC st_other: 0 st_shndx: 0 st_value: 0 st_size: 0
DEBUG:panopticon::loader: Import match printf: 0x601020 st_name: 16 GLOBAL FUNC st_other: 0 st_shndx: 0 st_value: 0 st_size: 0
DEBUG:panopticon::loader: Symbol: __libc_start_main @ 0x0: st_name: 23 GLOBAL FUNC st_other: 0 st_shndx: 0 st_value: 0 st_size: 0
DEBUG:panopticon::loader: Import match __libc_start_main: 0x600ff0 st_name: 23 GLOBAL FUNC st_other: 0 st_shndx: 0 st_value: 0 st_size: 0
DEBUG:panopticon::loader: Symbol: __gmon_start__ @ 0x0: st_name: 41 WEAK NOTYPE st_other: 0 st_shndx: 0 st_value: 0 st_size: 0
DEBUG:panopticon::loader: Imports: {
    6295576: "puts",
    6295584: "printf",
    6295536: "__libc_start_main"
}
```

This address is the same address used by the plt jmp stubs, so given the main function calling an internal function and two imports:

```asm
Dump of assembler code for function main:
   0x0000000000400564 <+0>:	push   %rbp
   0x0000000000400565 <+1>:	mov    %rsp,%rbp
   0x0000000000400568 <+4>:	mov    $0x0,%eax
   0x000000000040056d <+9>:	callq  0x400546 <internal>
   0x0000000000400572 <+14>:	mov    %eax,%esi
   0x0000000000400574 <+16>:	lea    0xb3(%rip),%rdi        # 0x40062e
   0x000000000040057b <+23>:	mov    $0x0,%eax
   0x0000000000400580 <+28>:	callq  0x400440 <printf@plt>
   0x0000000000400585 <+33>:	mov    $0x0,%eax
   0x000000000040058a <+38>:	callq  0x400551 <internal2>
   0x000000000040058f <+43>:	mov    $0x0,%eax
   0x0000000000400594 <+48>:	pop    %rbp
   0x0000000000400595 <+49>:	retq  
```

where, for example, the `callq 0x400440` (`printf@plt`) is a call to the jmp stub, e.g.:

```asm
(gdb) disass printf
Dump of assembler code for function printf@plt:
   0x0000000000400440 <+0>:	jmpq   *0x200bda(%rip)        # 0x601020
```

which accesses the import address of `printf` at (`0x601020`), which we record in `projects.imports` - `0x601020 -> printf`.

**NOTE** this still means jmp stub collapsing/recognition needs to be performed in order to print the `callq 0x400440` as `callq printf@plt`.

It also adds support for function pointers in dynamic relocations (like `__libc_start_main`), e.g.:

```asm
Dump of assembler code for function _start:
   0x0000000000400450 <+0>:	xor    %ebp,%ebp
   0x0000000000400452 <+2>:	mov    %rdx,%r9
   0x0000000000400455 <+5>:	pop    %rsi
   0x0000000000400456 <+6>:	mov    %rsp,%rdx
   0x0000000000400459 <+9>:	and    $0xfffffffffffffff0,%rsp
   0x000000000040045d <+13>:	push   %rax
   0x000000000040045e <+14>:	push   %rsp
   0x000000000040045f <+15>:	mov    $0x400610,%r8
   0x0000000000400466 <+22>:	mov    $0x4005a0,%rcx
   0x000000000040046d <+29>:	mov    $0x400564,%rdi
   0x0000000000400474 <+36>:	callq  *0x200b76(%rip)        # 0x600ff0
   0x000000000040047a <+42>:	hlt    
```

which is the `callq  *0x200b76(%rip)`, which we record as `0x600ff0 -> __libc_start_main` in the `projects.imports`

Note this is a global data dynamic relocatino, e.g.: `          600ff0 X86_64_GLOB_DAT __libc_start_main`

So basically, we have a structure now with the necessary information; the next step is to decide how to access the RIP relative values and resolve them into keys into the import table, and/or how to collapse the jmp stubs in a good way.

@flanfly @sphinxc0re let me know what you think, suggestions, concerns, etc.

### CODE REVIEW

**Comment by m4b** [comment]
Not sure why that load project test is failing, will look into this

**Comment by m4b** [comment]
Ah, the project deserialization will be different of course.  Not sure what the best approach here is - its easiest to just break backcompat obviously, but let me know what you have in mind

**Comment by flanfly** [comment]
Just update the ``save.panop`` file. What I do is I open ``tests/data/static`` and save it as ``save.panop``. The test is mostly acts as a canary for breaking changes to the on-disk format. There is not stability guarantee before 1.0. So no need to be particularly careful.

**Comment by m4b** [comment]
Hmm, i'm seeing `save.panop` currently a 74k file; but if i save `tests/data/static` its a 26MB file, this doesn't seem right?

**Comment by flanfly** [comment]
Should be ok. The save files are compressed.

**Comment by m4b** [comment]
Hmm, when i open `tests/data/sosse` i get splash screen with "No such file type", and drop down list of various microcontrollers.

All of them return result like: `qml: No such target 'mos6502'`

**Comment by flanfly** [comment]
Should work now when you select `atmega88`.

**Comment by flanfly** [comment]
LGTM

**Comment by sphinxc0re** [comment]
This looks really good but I would like to see a test for this. Could you please write one?

**Comment by m4b** [comment]
@sphinxc0re Yea for sure! It looks like we don't have a dynamically linked elf binary in our `tests/data`, should I add one? Or did you have some other test in mind?

**Comment by sphinxc0re** [comment]
No, just go ahead and add one 👍

**Comment by m4b** [comment]
don't merge this i just realized adding the symbol table values duplicates number of symbols :/ one sec

**Comment by m4b** [comment]
k last commit should fix, so re #254 we could at some point add tests like:

`assert_eq!(project.functions().len(), 15)`

which would prevent regressions like above, etc.

**Comment by m4b** [comment]
@sphinxc0re welllllp good thing you asked for a test; just realized that i'm loading 3 values into the import table (and checking against 3 in the elf test), but panopticon is (awesomely) discovering 5 jump stub functions, and similarly relocations show 5 jmp stubs, so something is wrong at the moment w.r.t. the import table loader, looking into it now

**Comment by m4b** [comment]
so the cause of this was that there are technically 2 (`puts`, `printf`) externally imported functions (global functions with no address in the dynamic table), but for the other 3 (`foo`, `bar`, `foobar`), which are functions exported by the lib itself, they are called by other functions call using the PLT (as if its an imported function).

Consequently, there are 5 entries in the PLT table, and hence 5 jump stubs/functions that panopticon discovers, but we currently only put global functions with no address in the import table. I belive this is resolved by removing the `is_import()` check.

So for `libfoo.so`, there will now be 6 functions in the import table, 5 functions, and one function pointer (`__cxa_finalize`):

```
imports: {2101288: "foobar", 2101272: "puts", 2101296: "printf", 2101304: "foo", 2101280: "bar", 2101240: "__cxa_finalize"}
```

**Comment by flanfly** [comment]
Ready?

File: lib/src/loader.rs [hunk 1]
```diff
@@ -105,19 +106,51 @@ fn load_elf(fd: &mut File, name: String) -> Result<(Project,Machine)> {
...
+    };
+
+    // add dynamic symbol information (non-strippable)
+    for sym in &binary.dynsyms {
+        add_sym(&mut prog, sym, &binary.dynstrtab);
+        if sym.is_function () {
+            let name = &binary.dynstrtab[sym.st_name];
+            if !resolve_import_address(&mut proj, &binary.pltrelocs, name) {
+                if !resolve_import_address(&mut proj, &binary.dynrelas, name) {
+                    resolve_import_address(&mut proj, &binary.dynrels,  name);
```
Comment (sphinxc0re) [inline lib/src/loader.rs]: I have one last question: Why didn't you define the method `resolve_import_address ` on `Project` itself?

Comment (m4b) [inline lib/src/loader.rs]: I suppose I could have, but its specific to ELF imports (e.g., the `&[Reloc]` argument), and can't be used with mach-o, or PE imports, and so I felt like hiding it in the ELF loader implementation details. Also doing so would expose goblin internals in the `Project` api, which doesn't seem necessary (i.e., I think its probably best for binary loading details to all be private functions called by the public and generic loader)

Speaking of mach-o, I've got a preliminary mach-o loader done on a local branch, and looks like imports are very similar (with a jump stub); I haven't thought about or looked at PE yet, but I suspect what we have here will work as a base.

So question is how we're going to use this import table to actually collapse the `callq`s into the names we have here (which is another issue, but wondering if we have anything in the works (I'm secretly hoping @flanfly has it all done already :laughing: ))?

Comment (sphinxc0re) [inline lib/src/loader.rs]: 👍 very good
